### ISSUE TITLE
Extending guide with DNS and letsencrypt

### ISSUE THREAD

**Issue by nabriis** [jupyterhub/zero-to-jupyterhub-k8s#15]
I think a guide for extending this zero to Jhub guide with a DNS server and letsencrypt would be essential.

If the DNS + letsencrypt set-up is made, it makes much more sense to detail different authenticators, since we really want SSL encryption before working with passwords etc.

Maybe include this in the default set-up, since the default currently uses no SSL, which is not recommended.

**Comment by yuvipanda** [comment]
Thanks for filing the issue!

I agree letsencrypt integration is the right thing to do.

**Comment by minrk** [comment]
We should definitely add letsencrypt to the walkthough!

Letsencrypt works right now, we just need to add it to the docs. FWIW, I believe this is the config necessary for letsencrypt to work in the helm chart as it is now:

```yaml
ingress:
  enabled: true
  annotations:
    kubernetes.io/ingress.class: nginx
  host: your-public-host.biz
  https:
    enabled: true
    type: "kube-lego"

proxy:
  service:
    type: ClusterIP
```

@yuvipanda is any of that superfluous? Could we reduce the necessary config to something more like:

```yaml
letsencrypt:
  domain: yourdomain.horse
  email: you@yourdomain.horse
```

which would trigger the ingress, ClusterIP, etc. config?

**Comment by cmoscardi** [comment]
https://hackmd.io/CYNgjKDsIEwLQGMCmBWALHNsVwIYlwwGYw0ZRdIAjXADhSA= might be more up to date?

**Comment by cmoscardi** [comment]
OK. Putting this here to document, @yuvipanda would love some help before JupyterCon next week. 

i'm currently running into this, I believe. the proxy service returns a 302 by default, so i think this breaks google cloud's health checking: 
https://github.com/jetstack/kube-lego/issues/18

You can go to https://train.thedataincubator.co/ and see that https has worked (I had to set the lego backend to be "gce", not "nginx"). However, the routing to the `proxy-public` service (which is also up on an external IP) is not correctly being routed through. Here's my ingress config:

```
spec:
  rules:
  - host: train.thedataincubator.co
    http:
      paths:
      - backend:
          serviceName: kube-lego-gce
          servicePort: 8080
        path: /.well-known/acme-challenge/*
      - backend:
          serviceName: proxy-public
          servicePort: 80
        path: /*
```


Long story short: I think that either `/` needs to return 200 or we need to fix the health check, and I'm not entirely clear how. Or I'm just doing something terribly wrong.

**Comment by Winterflower** [comment]
@cmoscardi just curious, when you ran
```
kubectl --namespace=<sth> get ingress
```
did the name of the kube-lego ingress first appear as kube-lego-nginx (as it does in my case)? Did you manually change it later to kube-lego-gce?

**Comment by cmoscardi** [comment]
hey, sorry for the delay @Winterflower - so, I'll be back with more shortly. But yes, I explicitly changed the service to `kube-lego-gce` I *think* because of a recommendation I saw somewhere. It was nginx by default (and something really weird was happening with that, too).

**Comment by Winterflower** [comment]
hihi @cmoscardi !
super - thanks very much for those screencaps. I think in the Google Cloud console screenshot, one of the kube-lego-nginx services if of type Ingress (Resource) and the other ClusterIP (this is for the kube-lego-nginx pod ). I think kube-lego pod creates an Ingress resource for the ACME challenge by default, and calls it (confusingly) kube-lego-nginx, which would explain why you have two entries. 
I went with the default kube-lego-gce and my console also has two entries - one called kube-lego-gce (for the pod) and another to kube-lego-nginx (the type is Ingress).

**Comment by cmoscardi** [comment]
So, after enough fiddling with it, I did get it to work! In particular, I deleted the "default https check" under the [health checks page](https://console.cloud.google.com/compute/healthChecks) and things started working. 

How does this work? We just don't know. But I now might be able to fashion a PR out of this (for GKE people, at least).

**Comment by cmoscardi** [comment]
Right. Now, on the latest master, I'm encountering a new thing. 

From chrome:
```
This site can’t be reached
t5.thedataincubator.co unexpectedly closed the connection.
```

This is with the "gce" ingress above, health check set appropriately (and apparently everything is healthy). Let's encrypt has successfully run (according to the logs). There's probably some firewall rule I need to set?

**Comment by cmoscardi** [comment]
As a point of good practice, I'll remark that I did get gce ingress working. 

If you're reading this and want to also make it work, feel free to reply here and I'll get back to you ASAP with my instructions. It's a rather intricate set of steps.

It doesn't seem to be of general interest, though (there's a motion to get [nginx](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/212) working)

**Comment by minrk** [comment]
#229 adds support for SSL with letsencrypt with:

```yaml
proxy:
  hosts:
    - <your-host>
  letsencrypt:
    contactEmail: <your-email>
```
